File: ruby/arraySearchKSum.rb
#!/usr/bin/ruby
# frozen_string_literal:false

# a Class to find the first numbers in an array that sum K, an Integer number
class KSumSearcher
  def initialize(input_array)
    @the_array = input_array
    @registry = {}
    @combinations = {}
    @possible_index = []
  end

  def combinations_keys_iteration(a_number, index)
    temp_combinations = {}
    @combinations.each_key do |saved_combination|
      return @combinations[saved_combination] if a_number + saved_combination == @k_number

      if (a_number + saved_combination).abs < @k_number.abs
        temp_combinations[a_number + saved_combination] = [@combinations[saved_combination], index].flatten
      end
    end
    @combinations.merge!(temp_combinations) { |_ky, oldval, _newval| oldval }
    []
  end

  def registry_keys_iteration(a_number, index)
    @registry.each_key do |saved|
      return [@registry[saved].first] if a_number + saved == @k_number

      if (a_number + saved).abs < @k_number.abs && @combinations[a_number + saved].nil?
        @combinations[a_number + saved] = [@registry[saved].first, index]
      end
    end
    []
  end

  def no_success_with_registred
    @possible_index[1].empty? || (!@possible_index[0].empty? && @possible_index[1][0] > @possible_index[0].first)
  end

  def search_solution(index)
    return [@possible_index[0], index].flatten if no_success_with_registred

    [@possible_index[1], index].flatten
  end

  def process_number(a_number, index)
    @possible_index.append(combinations_keys_iteration(a_number, index))
    @possible_index.append(registry_keys_iteration(a_number, index))

    return search_solution(index) unless @possible_index.flatten.empty?

    @possible_index.clear
    nil
  end

  def search(k_number)
    @k_number = k_number
    @the_array.each_with_index do |a_number, index|
      return index unless @k_number - a_number != 0

      result = process_number(a_number, index)
      return result unless result.nil?

      @registry[a_number] = [index] if @registry[a_number].nil?
    end
    'no se encontro resultado'
  end
end

require 'json'
# receive an array like [1,2,3], and a number, returns the expected array of indexes
p KSumSearcher.new(JSON.parse(ARGV[0])).search(ARGV[1].to_i)

File: java/ArraySearchKSum.java
import java.util.*;

import static java.lang.Math.abs;
import static java.lang.System.out;


// a Class to find the first numbers in an array that sum K, an Integer number
public class ArraySearchKSum {
    private static final List<List<Integer>> possible_index = new ArrayList<>();
    private static final Map<Integer, List<Integer>> combinations = new HashMap<>();
    private static final Map<Integer, List<Integer>> registry = new HashMap<>();

    private static int kNumber;
    @SafeVarargs
    private static <T> List<T> mutateAdd(final List<T> al, final T... ts) {
        final List<T> bl = new ArrayList<>(al);
        Collections.addAll(bl, ts);
        return List.copyOf(bl);
    }

    static List<Integer> combinationsKeysIteration(Integer aNumber, int index) {
        Map<Integer, List<Integer>> tempCombinations = new HashMap<>();
        for (Map.Entry<Integer, List<Integer>> combination : combinations.entrySet()) {
            Integer combinationKey = combination.getKey();
            List<Integer> value = combination.getValue();
            if (aNumber + combinationKey == kNumber) return value;
            if (abs(aNumber + combinationKey) < abs(kNumber)) {
                tempCombinations.put(aNumber + combinationKey, mutateAdd(value, index));
            }
        }
        tempCombinations.forEach((key, value) ->
                combinations.merge(key, value, (v1, v2) -> v1));
        return new ArrayList<>();
    }

    static List<Integer> registryKeysIteration(Integer aNumber, int index) {
        int savedKey;
        for (Map.Entry<Integer,List<Integer>> entry : registry.entrySet()) {
            savedKey= entry.getKey();
            if (aNumber + savedKey == kNumber) return List.of(entry.getValue().get(0));
            if (abs(aNumber + savedKey) < abs(kNumber) && combinations.get(aNumber + savedKey) == null) {
                combinations.put(aNumber + savedKey, List.of(entry.getValue().get(0), index));
            }
        }
        return Collections.emptyList();
    }

    static boolean noSuccessWithRegistred() {
        return possible_index.get(1).isEmpty() || (!possible_index.get(0).isEmpty() && possible_index.get(1).get(0) > possible_index.get(0).get(0));
    }

    static List<Integer> searchSolution(Integer index) {
        return mutateAdd(possible_index.get(noSuccessWithRegistred() ? 0 : 1),index);
    }

    static List<Integer> processNumber(int aNumber, int index) {
        possible_index.add(combinationsKeysIteration(aNumber, index));
        possible_index.add(registryKeysIteration(aNumber, index));
        if ((possible_index.stream()
                .mapToLong(List::size).sum() != 0)) return searchSolution(index);
        possible_index.clear();
        return null;
    }
    //receive an array like '1 2 3', and a number, returns the expected array of indexes
    public static void main(String[] args) {
        kNumber = Integer.parseInt(args[0]);
        final String[] num = args[1].split(" ");
        final List<Integer> numbers = new ArrayList<>();
        for (final String cur : num) {
            numbers.add(Integer.valueOf(cur));
        }
        Integer[] theArray = numbers.toArray(new Integer[0]);
        for (int index = 0; index < theArray.length; index++) {
            int aNumber = theArray[index];
            if (kNumber - aNumber == 0) {
                out.println(index);
                return;
            }
            var result = processNumber(aNumber, index);
            if (result != null) {
                out.println(result);
                return;
            }
            int finalIndex = index;
            registry.computeIfAbsent(aNumber, k -> List.of(finalIndex));
        }
        out.println("no se encontro resultado");
    }
}